src/extension/vscode/server.ts

```diff
@@ -0,0 +1,110 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import type { WebSocketServer, RawData, WebSocket } from "ws";
+import type { MessageAdapter } from "../messageAdapters";
+import { createRpcClient, createRpcHandler } from "../rpc";
+import type { Actor } from "../actor";
+import { createActor } from "../actor";
+import type { ApolloClientInfo } from "../../types";
+import { createId } from "../../utils/createId";
+import allErrorCodes from "../../../all-clients/errorcodes.json";
+import { restoreErrorCodes } from "../../../all-clients/restore-errorcodes.mjs";
+
+// copied into this file to ease type bundling for now
+export interface PublicMessageAdapter {
+  addListener: (listener: (message: unknown) => void) => () => void;
+  postMessage: (message: unknown) => void;
+}
+
+export function runServer(
+  server: WebSocketServer,
+  adapter: PublicMessageAdapter
+) {
+  const clients = new Map<
+    string,
+    {
+      rpcClient: ReturnType<typeof createRpcClient>;
+      actor: Actor;
+      info: ApolloClientInfo;
+    }
+  >();
+
+  const messageAdapter = adapter as MessageAdapter;
+  const tab = createActor(messageAdapter);
+  const handleRpc = createRpcHandler(messageAdapter);
+
+  handleRpc("getClients", () =>
+    Array.from(clients.values()).map(({ info }) => info)
+  );
+  handleRpc("getClient", (id) => clients.get(id)!.info);
+  handleRpc("getQueries", (id) =>
+    clients.get(id)!.rpcClient.request("getQueries", id)
+  );
+  handleRpc("getMutations", (id) =>
+    clients.get(id)!.rpcClient.request("getMutations", id)
+  );
+  handleRpc("getCache", (id) =>
+    clients.get(id)!.rpcClient.request("getCache", id)
+  );
+
+  server.on("connection", function connection(ws) {
+    const id = createId();
+    const wsAdapter = createSocketMessageAdapter(ws);
+    const wsRpcClient = createRpcClient(wsAdapter);
+    const wsRpcHandler = createRpcHandler(wsAdapter);
+    const wsActor = createActor(wsAdapter);
+    wsActor.on("registerClient", ({ payload }) => {
+      payload.id = id;
```

Comment:
Ah ok I see this is set here instead of in `client.ts`, which I missed when adding the other comment. 

I'm curious why you decided to go this route of creating a handler/rpc client per apollo client instance? If you're able to use a single client/rpc handler, you would then be able to take advantage of the [message bridge](https://github.com/apollographql/apollo-client-devtools/blob/a413e0b8fe2c3ccab7225950439bbf6e9dccab62/src/extension/messageAdapters.ts#L104-L124) since this script would essentially just be the middle man between the client script and devtools.

I'd recommend taking a stab at that, otherwise we'll need to edit this file anytime we want to add a new message from devtools to the client (such as the memory panel that will be added soon). I used to do this in `tab.ts` and it there were definitely times that I scratched my head wondering why something wasn't working only to realize I forgot to add the message forwarding between the two. The message bridge definitely helps out since its a bit more automatic.

Reply:
The message bridge would work if we had the guarantee of a single connection, but you can totally connect two different processes to one server (e.g. connect an iOS emulator and an Android emulator at the same time). So either at that point I send everything everywhere, or I have to keep track of who is who in the server - I went for the latter.

Also, we could in the future add some "snapshot" functionality here inside the server that would work by creating a "virtual client" that would be fully controlled by the server.

Reply:
Creating the `id` in`registerClient` now, I just hope there won't be any collisions in low-entropy environments

Reply:
I don't know that I'm that worried to be honest. At most we'll have _maybe_ 10s of clients connected at once, at least in the 99% of cases. If we get reports of collisions between connected clients because we start getting tons of them, we can always revisit.